pandas/core/internals/blocks.py

```diff
@@ -1403,17 +1401,17 @@ def _unstack(self, unstacker, new_columns, fill_value, value_columns):
         mask : array_like of bool
             The mask of columns of `blocks` we should keep.
         """
-        new_items = unstacker.get_new_columns(value_columns)
-        new_placement = new_columns.get_indexer(new_items)
         new_values, mask = unstacker.get_new_values(
             self.values.T, fill_value=fill_value
         )
 
         mask = mask.any(0)
+        # TODO: in all tests we have mask.all(); can we rely on that?
+
         new_values = new_values.T[mask]
         new_placement = new_placement[mask]
 
-        blocks = [make_block(new_values, placement=new_placement)]
+        blocks = [self.make_block_same_class(new_values, placement=new_placement)]
```

Comment:
@jbrockmendel I think it's just this line that caused the regression in #37115 and that we don't need to revert the complete PR.

Reply:
what kind of block(s) do you end up with instead?

Reply:
corrupted? an IntBlock with dtype float64

```
>>> pd.__version__
'1.2.0.dev0+1432.g5fdf642368'
>>>
>>> df1 = pd.DataFrame(
...     {
...         "a": ["A", "A", "B"],
...         "b": ["ca", "cb", "cb"],
...         "v": [10] * 3,
...     }
... )
>>> df1 = df1.set_index(["a", "b"])
>>> df1["is_"] = 1
>>> df1
       v  is_
a b
A ca  10    1
  cb  10    1
B cb  10    1
>>>
>>> df1._data
BlockManager
Items: Index(['v', 'is_'], dtype='object')
Axis 1: MultiIndex([('A', 'ca'),
            ('A', 'cb'),
            ('B', 'cb')],
           names=['a', 'b'])
IntBlock: slice(0, 1, 1), 1 x 3, dtype: int64
IntBlock: slice(1, 2, 1), 1 x 3, dtype: int64
>>>
>>> df2 = df1.unstack("b")
>>> df2
      v        is_
b    ca    cb   ca   cb
a
A  10.0  10.0  1.0  1.0
B   NaN  10.0  NaN  1.0
>>>
>>> df2._data
BlockManager
Items: MultiIndex([(  'v', 'ca'),
            (  'v', 'cb'),
            ('is_', 'ca'),
            ('is_', 'cb')],
           names=[None, 'b'])
Axis 1: Index(['A', 'B'], dtype='object', name='a')
IntBlock: slice(0, 2, 1), 2 x 2, dtype: float64
IntBlock: slice(2, 4, 1), 2 x 2, dtype: float64
>>>
```

Reply:
run the test suite with just this change reverted and all tests pass. will put up a PR shortly.